src/osx/cocoa/dataview.mm

```diff
@@ -2892,34 +2894,67 @@ void UpdateWithRow(int row)
         //else: can't change font if the cell doesn't have any
     }
 
-    if ( attr.HasColour() && [cell backgroundStyle] == NSBackgroundStyleLight )
+    // We don't apply the text or background colours if the cell is selected.
+    if ( [cell backgroundStyle] == NSBackgroundStyleLight )
     {
-        // we can set font for any cell but only NSTextFieldCell provides
-        // a method for setting text colour so check that this method is
-        // available before using it
-        if ( [cell respondsToSelector:@selector(setTextColor:)] &&
-                [cell respondsToSelector:@selector(textColor)] )
+        if ( attr.HasColour() )
         {
-            if ( !data->GetOriginalTextColour() )
+            // we can set font for any cell but only NSTextFieldCell provides
+            // a method for setting text colour so check that this method is
+            // available before using it
+            if ( [cell respondsToSelector:@selector(setTextColor:)] &&
+                    [cell respondsToSelector:@selector(textColor)] )
             {
-                // the cast to (untyped) id is safe because of the check above
-                data->SaveOriginalTextColour([(id)cell textColor]);
+                if ( !data->GetOriginalTextColour() )
+                {
+                    // the cast to (untyped) id is safe because of the check above
+                    data->SaveOriginalTextColour([(id)cell textColor]);
+                }
+
+                colText = attr.GetColour().OSXGetNSColor();
             }
+        }
 
-            colText = attr.GetColour().OSXGetNSColor();
+        if ( attr.HasBackgroundColour() )
+        {
+            // Use the same logic as the text colour check above
+            if ( [cell respondsToSelector:@selector(setBackgroundColor:)] &&
+                    [cell respondsToSelector:@selector(backgroundColor)] )
+            {
+                if ( !data->GetOriginalBackgroundColour() )
+                    data->SaveOriginalTextColour([(id)cell backgroundColor]);
+
+                colBack = attr.GetBackgroundColour().OSXGetNSColor();
+            }
         }
     }
 
+
     if ( !font )
         font = data->GetOriginalFont();
     if ( !colText )
         colText = data->GetOriginalTextColour();
+    if ( !colBack )
+        colBack = data->GetOriginalBackgroundColour();
 
     if ( font )
         [cell setFont:font];
 
     if ( colText )
         [(id)cell setTextColor:colText];
+
+    if ( [cell respondsToSelector:@selector(setDrawsBackground:)] )
+    {
+        if ( colBack )
+        {
+            [(id)cell setDrawsBackground:true];
+            [(id)cell setBackgroundColor:colBack];
+        }
+        else
+        {
+            [(id)cell setDrawsBackground:false];
```

Comment:
Is this branch actually ever taken? It looks like `colBack` will always be set to something, because of fallback to the original bg colour.

Reply:
Yes, I think this branch could end up being taken. The saving of the original background color only happens if the current item attribute has a background color set. By default, the original color field inside the native data is `NULL`, so it could take this branch if it draws an item without a background color before getting one with a background color.

Perhaps it would actually be better to make this call when `HasBackgroundColor` is false though, since we won't have a background color to draw then.